Cargo boxcutter

Shadyyy66 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This PR adds a very simple boxcutter to the game, it does as much damage as an unsharpened kitchen knife, makes a fun little noise when being unfolded and folded and can very quickly unwrap wrapped packages and return the wrapping paper that was used to wrap without making a mess that needs to be cleaned up, it even has a custom sound for unwrapping!

This PR also adds a new proc for how parcels or wrapped objects in sortingmachinery.dm are opened by adding the "tear_open" proc, if it's set to TRUE, it takes time to unwrap and it leaves a mess, but if it's set to FALSE, it is instantly unwrapped and drops one (1) package wrapping and plays the custom box cutting sound clip I added.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Honestly, I just think it would be fun for cargo to have a tool of their own that isn't just a scanner or whatever they can make with the autolathe. Plus, as a frequent janitor player, it leaves less messes in cargo that janitors usually don't have access to anyway, and wastes less package wrapping!

Not only that, but it feels like cargo itself can be lacking in department-specific weapons to use. Every other department has some kind of weapon tied to it, like service and its knives and hatchets, or medical with its scalpels and saws.
Basically, it feels like cargo isn't living up to the same potential as the other departments, and mining might as well be its own department with how separate they are.

And honestly, cargo techs deserve some kind of weapon to defend themselves with, feels like there's always something going down in cargo.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Added boxcutter to cargo.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
![Untitled-3](https://user-images.githubusercontent.com/114319683/192625769-ddd1607c-142a-43db-85cc-198a6078ff26.gif)

Shadyyy66 committed:
Create boxcutter.dm

Shadyyy66 committed:
Update boxcutter.dm

Shadyyy66 committed:
ADDED BOXCUTTER

HUH?? WHAT??

Shadyyy66 committed:
Added parcel interaction

yeah

Shadyyy66 committed:
Added, uh, stuff

can unfold now, plus honks when used on parcels

Shadyyy66 committed:
Boxcutter sprites

added sprites for the boxcutter

Shadyyy66 committed:
Boxcutter icon names

changed boxcutter icon names

Shadyyy66 committed:
Fixed Boxcutter Sprites

Boxcutter sprites now appear in-game and are also now functional in unwrapping parcels

Shadyyy66 committed:
boxcutter sound testing

yah

Shadyyy66 committed:
Added state recognition + sound

parcels now recognize the difference between an active and inactive boxcutter, also added a new sound for when the boxcutter is activated

Shadyyy66 committed:
Updated boxcutter inhand sprites

yah

Shadyyy66 committed:
Boxcutter inhand sprites updated

again

JohnFulpWillard commented:
![image](https://user-images.githubusercontent.com/53777086/192629752-391a680d-7cfb-4b57-8a96-fe37e55d8b94.png)

Mothblocks commented on code/game/objects/items/boxcutter.dm:
```diff
@@ -0,0 +1,42 @@
+	. = ..()
+	AddElement(/datum/element/update_icon_updates_onmob, ITEM_SLOT_HANDS)
+	AddComponent(/datum/component/butchering, \
+	speed = 7 SECONDS, \
+	effectiveness = 100, \
```
Indent up by one

Mothblocks commented on code/game/objects/items/boxcutter.dm:
```diff
@@ -0,0 +1,42 @@
+		hitsound_on = 'sound/weapons/bladeslice.ogg', \
+		w_class_on = WEIGHT_CLASS_NORMAL, \
+		attack_verb_continuous_on = list("cuts", "stabs", "slashes"), \
+		attack_verb_simple_on = list("cut", "stab", "slash"))
```
```suggestion
		attack_verb_simple_on = list("cut", "stab", "slash"), \
	)
```

Mothblocks commented on code/game/objects/items/boxcutter.dm:
```diff
@@ -0,0 +1,42 @@
+/obj/item/boxcutter/proc/on_transform()
+	SIGNAL_HANDLER
+	playsound(src, on_sound, 50)
+	return COMPONENT_NO_DEFAULT_MESSAGE
```
How noticable is the in-hand sprite being changed? Can you show a picture with switch blade extended and not? The message should only be silenced if looking at the center of your screen makes it obvious that you've activated it--we don't want people to have to move their eyes from where they're looking/rely on the sound alone.

Mothblocks replied:
ill take your word for it

JohnFulpWillard commented on code/game/objects/items/boxcutter.dm:
```diff
@@ -0,0 +1,42 @@
+
+/obj/item/boxcutter/proc/on_transform()
+	SIGNAL_HANDLER
+	playsound(src, on_sound, 50)
```
You can instead make this part of the transforming component, since it's currently hardcoded to batonextend.

JohnFulpWillard commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
 		update_appearance()
 
+	if(istype(item, /obj/item/boxcutter))
+		if( (item.icon_state == "boxcutter_on"))
```
The boxcutter should have an 'on' var instead, looking at the icon state is a bad idea,

Mothblocks replied:
Have a var that updates instead of checking icon_state directly.

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
 			wrapped_item.AddComponent(/datum/component/pricetag, sticker.payments_acc, sticker.cut_multiplier)
 		update_appearance()
 
+	if(istype(item, /obj/item/boxcutter))
```
```suggestion
	else if(istype(item, /obj/item/boxcutter))
```

Otherwise `..()` will hit for all the other items.

Also I kind of hate that this entire proc looks like this but probably not worth it for you to change right now.

JohnFulpWillard commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
+			post_unwrap_contents(user, rip_open = FALSE)
+			return
+		else
+			balloon_alert(user, span_warning("The boxcutter isn't active!"))
```
```suggestion
			balloon_alert(user, span_warning("open [item] first!"))
```

Mothblocks replied:
this is wrong because it's a balloon alert and balloon alerts should not have spans and should be kept as small as possible

JohnFulpWillard replied:
problem is 'open it first' or something makes you question if it means the boxcutter or the wrapper

Mothblocks replied:
The suggestion I made was to change it to "prime the boxcutter", you wouldn't tell someone to "open" a switchblade anyway

JohnFulpWillard commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
 
+	if(istype(item, /obj/item/boxcutter))
+		if( (item.icon_state == "boxcutter_on"))
+			unwrap_contents(user)
```
can you make this use the full unwrap code instead of shortcut to the end? It still takes time to open packages with a boxcutter

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
+			post_unwrap_contents(user, rip_open = FALSE)
+			return
+		else
+			balloon_alert(user, span_warning("The boxcutter isn't active!"))
```
```suggestion
			balloon_alert(user, "prime the boxcutter!")
```

Balloon alerts are not styled the same way as chat messages, and the wording on this wasn't very boxcutter-y. Feel free to replace prime with arm or ready.

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -185,6 +188,15 @@
+		if( (item.icon_state == "boxcutter_on"))
+			unwrap_contents(user)
+			post_unwrap_contents(user, rip_open = FALSE)
+			return
```
Remove both these since they always return anyway.

Shadyyy66 committed:
Boxcutter spawns

boxcutter spawns in cargo tech backpacks and can be made in the cargo techfab and autolathe

Shadyyy66 committed:
Update code/game/objects/items/boxcutter.dm

uhh adds a , \ to the end of attack_verb_simple_on

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

optimumtact commented:
![maintainers add this](https://file.house/PjoY.gif)

Shadyyy66 committed:
Apply suggestions from code review

adds an else statement to the start of if(istype(item, /obj/item/boxcutter))

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/recycling/sortingmachinery.dm

change of wording on balloon_alert

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 committed:
wording and indents updated

Shadyyy66 committed:
Merge branch 'cargo-boxcutter' of https://github.com/Shadyyy66/tgstation into cargo-boxcutter

Farquaar commented:
> You could probably do to buff its damage a bit. Unsharpened knives are at 10, cleavers and miner's knives are at 15. It's a propper tool and should be more effective than a shiv. 10 is a safe number and on par with the lesser knife tools. Should also have a wound bonus and maybe `tool_behaviour = TOOL_KNIFE`

I mean, it's a box cutter; I think the damage value seems fair for what it is. It definitely should deal less damage than a kitchen knife.

Shadowflame909 commented:
a damage buff to equal all other knives cool for making it useful for other situations in this sandbox. But the tool_behavior knife property would be good during the days I want to cut ian into giblets with the boxcutter instead of a knife.

It opens up options!

Shadyyy66 committed:
Boxcutter timer, optimization, balancing

Mothblocks commented on code/modules/jobs/job_types/cargo_technician.dm:
```diff
@@ -39,6 +39,9 @@
 
+	backpack_contents = list(
+		/obj/item/boxcutter = 1,
+		)
```
```suggestion
	)
```

I know they're all like this but they're all wrong

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -12,9 +12,9 @@
  * Initial check if manually unwrapping
  */
-/obj/item/delivery/proc/attempt_pre_unwrap_contents(mob/user)
+/obj/item/delivery/proc/attempt_pre_unwrap_contents(mob/user, time = 15)
```
```suggestion
/obj/item/delivery/proc/attempt_pre_unwrap_contents(mob/user, time = 1.5 SECONDS)
```

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -189,13 +189,15 @@
+			if(!attempt_pre_unwrap_contents(user, time = 5))
+				return
+			unwrap_contents()
+			to_chat(user, span_notice("You start to cut open package..."))
```
```suggestion
			balloon_alert(user, "cutting open package...")
```

Mothblocks commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -189,13 +189,15 @@
-			unwrap_contents(user)
+		var/obj/item/boxcutter/boxcutter_item = item
+		if(boxcutter_item.on)
+			if(!attempt_pre_unwrap_contents(user, time = 5))
```
```suggestion
			if(!attempt_pre_unwrap_contents(user, time = 0.5 SECONDS))
```

Shadyyy66 replied:
i dont know if i want to change this to seconds, is there even any code that uses it? im worried its gonna break

itseasytosee replied:
Don't worry man, seconds is literally just defined as `* 10` its just for readability.

Mothblocks commented on code/game/objects/items/boxcutter.dm:
```diff
@@ -13,19 +13,21 @@
 	resistance_flags = FIRE_PROOF
 	force = 0
 	var/start_extended = FALSE
+	var/on = FALSE
```
```suggestion
	/// Whether or not the boxcutter has been readied
	var/on = FALSE
```

Shadyyy66 committed:
Update code/game/objects/items/boxcutter.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/recycling/sortingmachinery.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/jobs/job_types/cargo_technician.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 reviewed (commented):
just some formatting changes

Shadyyy66 committed:
Update code/modules/recycling/sortingmachinery.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/recycling/sortingmachinery.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>
